### PR TITLE
[UX] Add commit hash

### PR DESCRIPTION
**Pull request by Michaelvll**
Make it possible to get commit hash of the current SkyPilot for both installed from pypi and source code.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] `pip install -e .; cd ~; sky --commit; sky launch -c test --cpus 2+; sky spot launch -n test-aws --cpus 2 --cloud aws`
  - [x] `pip install .; cd ~; sky --commit; sky launch -c test --cpus 2+ echo hi; sky down sky-spot-controller-xxx`
  - [x] Upload to test pypi with the building script in nightly build script, `pip install -i https://test.pypi.org/simple/ skypilot==1.0.0.dev20231020; sky --commit`
  - [x] `pip install -e .; sky launch -c test-commit --cpus 2 echo hi; ssh test-commit; sky -c`
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/__init__.py [hunk 1]
```diff
@@ -1,12 +1,35 @@
...
+    except Exception:  # pylint: disable=broad-except
+        return _SKYPILOT_COMMIT_SHA
+
+
+__commit__ = get_git_commit()
 __version__ = '1.0.0-dev0'
 __root_dir__ = os.path.dirname(os.path.abspath(__file__))
 
 # Keep this order to avoid cyclic imports
+# pylint: disable=wrong-import-position
```
Comment (cblmemo) [inline sky/__init__.py:32]: Why add this? Seems like not relevant to this PR..?

**Review by cblmemo — COMMENTED** [review]
Thanks for adding this! I'm a little confused about the default value `{{SKYPILOT_COMMIT_SHA}}`. Where is it replaced by the commit hash? I tried searching in the repository but did not find anything helpful..

**Comment by Michaelvll** [comment]
> Thanks for adding this! I'm a little confused about the default value `{{SKYPILOT_COMMIT_SHA}}`. Where is it replaced by the commit hash? I tried searching in the repository but did not find anything helpful..

We did not do it in our code, but manually replace it for each release before this PR, and in our CI https://github.com/skypilot-org/skypilot/blob/d38b4b4163e4d5f4ef98895e30b4b5c04fd23dde/.github/workflows/pypi-nightly-build.yml

File: sky/__init__.py (hunk 1 above)
Comment (Michaelvll) [reply sky/__init__.py:32]: Previously, the `pylint` did not check this file as the file was not changed for a long time before, and it has `wrong-import-position` warning now.

**Review by cblmemo — APPROVED** [review]
Thanks for the explanation! LGTM 🫡

**Comment by Michaelvll** [comment]
We fixed the logic for `setup` and `__init__` to make sure `__init__` reverts to original content, so that it can correctly handle `python -m build`. PTAL @cblmemo.

File: sky/setup_files/setup.py [hunk 2]
```diff
@@ -13,32 +13,91 @@
...
 
 
+def get_commit_hash():
+    with open(INIT_FILE_PATH, 'r') as fp:
+        commit_match = re.search(r'^_SKYPILOT_COMMIT_SHA = [\'"]([^\'"]*)[\'"]',
+                                 fp.read(), re.M)
+        if commit_match:
+            commit_hash = commit_match.group(1)
+        else:
+            raise RuntimeError('Unable to find commit string.')
```
Comment (cblmemo) [inline sky/setup_files/setup.py:52]: Can we reuse the `find_version` function? Seems like a lot of code is duplicated

File: sky/setup_files/setup.py [hunk 3]
```diff
@@ -13,32 +13,91 @@
...
+            cwd=cwd,
+            universal_newlines=True,
+            stderr=subprocess.DEVNULL).strip()
+        changes = subprocess.check_output(['git', 'status', '--porcelain'],
+                                          cwd=cwd,
+                                          universal_newlines=True,
+                                          stderr=subprocess.DEVNULL).strip()
+        if changes:
+            commit_hash += '-dirty'
+        return commit_hash
```
Comment (cblmemo) [inline sky/setup_files/setup.py:69]: Add a comment that says this should be aligned with implementation in `sky/__init__.py`?

File: sky/setup_files/setup.py [hunk 4]
```diff
@@ -13,32 +13,91 @@
...
+        fp.write(content)
+
+
+def revert_commit_hash():
+    try:
+        if original_init_content is not None:
+            with open(INIT_FILE_PATH, 'w') as fp:
+                fp.write(original_init_content)
+        else:
+            subprocess.check_call(['git', 'checkout', INIT_FILE_PATH],
```
Comment (cblmemo) [inline sky/setup_files/setup.py:94]: What if the user has made some changed to the init file? IIUC this command will discard these changes? (Though arguably it is a rare case)

File: sky/setup_files/setup.py [hunk 5]
```diff
@@ -13,32 +13,91 @@
...
+                                  cwd=ROOT_DIR,
+                                  stderr=subprocess.DEVNULL)
+    except subprocess.CalledProcessError:
+        pass
```
Comment (cblmemo) [inline sky/setup_files/setup.py:98]: raise some error to indicate the revert failed?

**Review by cblmemo — COMMENTED** [review]
Thanks for the fix! Left several comments 😉

File: sky/setup_files/setup.py (hunk 2 above)
Comment (Michaelvll) [reply sky/setup_files/setup.py:52]: It is short enough, maybe it is ok to duplicate a bit? Making it another function may take longer lines.

Comment (cblmemo) [reply sky/setup_files/setup.py:52]: Make sense. Agreed 👍

File: sky/setup_files/setup.py (hunk 4 above)
Comment (Michaelvll) [reply sky/setup_files/setup.py]: Good point! I removed the part for `git checkout` as we already reverts the content with our own `original_content`.

**Pull request merged**